File: services/transaction/royalty-money-math.ts
import { Decimal } from 'decimal.js';

/**
 * Money math for the Lyrica royalty receipt loop -- see
 * spec/SPEC-royalty-loop-v1.md §4. Owned by the transaction service
 * (which owns ledger posting), not the gateway -- the gateway only
 * relays what this module (via royalty-service-core.ts) computes.
 *
 * The pool (event.amount.value) belongs to the creator(s) in full.
 * There is no platform fee in this loop.
 */

export interface Split {
  ownerId: string;
  bps: number;
}

export interface Payout {
  ownerId: string;
  amount: Decimal;
}

const CENT = new Decimal('0.01');
const BPS_DENOMINATOR = new Decimal(10000);

export function distributeSplits(pool: Decimal, splits: Split[]): Payout[] {
  const totalBps = splits.reduce((sum, s) => sum + s.bps, 0);
  if (totalBps !== 10000) {
    throw new Error(`splits[].bps must sum to exactly 10000, got ${totalBps}`);
  }

  const raw = splits.map((s) => ({
    ownerId: s.ownerId,
    bps: s.bps,
    raw: pool.mul(s.bps).div(BPS_DENOMINATOR),
  }));

  const floored = new Map<string, Decimal>();
  const remainder = new Map<string, Decimal>();
  const bpsByOwner = new Map<string, number>();
  for (const r of raw) {
    const f = r.raw.toDecimalPlaces(2, Decimal.ROUND_DOWN);
    floored.set(r.ownerId, f);
    remainder.set(r.ownerId, r.raw.minus(f));
    bpsByOwner.set(r.ownerId, r.bps);
  }

  const poolCents = pool.div(CENT).toDecimalPlaces(0, Decimal.ROUND_DOWN);
  let floorCentsTotal = new Decimal(0);
  for (const f of floored.values()) {
    floorCentsTotal = floorCentsTotal.plus(f.div(CENT));
  }
  const leftoverCents = poolCents.minus(floorCentsTotal).toNumber();

  const ordering = splits
    .map((s) => s.ownerId)
    .sort((a, b) => {
      const remainderCompare = remainder.get(b)!.comparedTo(remainder.get(a)!); // descending remainder
      if (remainderCompare !== 0) return remainderCompare;
      const bpsCompare = bpsByOwner.get(b)! - bpsByOwner.get(a)!; // descending bps
      if (bpsCompare !== 0) return bpsCompare;
      return a < b ? -1 : a > b ? 1 : 0; // ascending owner_id
    });

  for (let i = 0; i < leftoverCents; i++) {
    const ownerId = ordering[i];
    floored.set(ownerId, floored.get(ownerId)!.plus(CENT));
  }

  return splits.map((s) => ({ ownerId: s.ownerId, amount: floored.get(s.ownerId)! }));
}

export interface RoyaltyBreakdown {
  gross: Decimal;
  platformFee: Decimal;
  net: Decimal;
  payouts: Payout[];
}

export function computeRoyalty(gross: Decimal, splits: Split[]): RoyaltyBreakdown {
  const pool = gross.toDecimalPlaces(2);
  const payouts = distributeSplits(pool, splits);

  const totalPayout = payouts.reduce((sum, p) => sum.plus(p.amount), new Decimal(0));
  if (!totalPayout.equals(pool)) {
    throw new Error(`balanced-journal invariant violated: payouts(${totalPayout}) != pool(${pool})`);
  }

  return { gross, platformFee: new Decimal('0.00'), net: gross, payouts };
}

File: services/gateway/royalty_money_math.py
"""
Money math for the Lyrica royalty receipt loop — see spec/SPEC-royalty-loop-v1.md §4.

Pure functions only: no I/O, no FastAPI, no gateway/orchestrator imports.
All arithmetic happens in Decimal; amounts are formatted as fixed-point
strings with exactly 4 decimal places to match the receipt schema and
the v1 monolith's BigInt ledger units (value * 10000).

`amount.value` on the incoming event IS the creator royalty pool, in
full. Archisynapse deducts nothing from it — there is no platform fee
in this loop. Empire's separate 70/30 platform-revenue-share policy
governs a different pricing contract and must not be applied here
unless a distinct, founder-approved contract says otherwise.
"""

from decimal import Decimal, ROUND_DOWN
from typing import Iterable, TypedDict

CENT = Decimal("0.01")
LEDGER_PLACES = Decimal("0.0001")
BPS_DENOMINATOR = Decimal(10000)


class Split(TypedDict):
    owner_id: str
    bps: int


class Payout(TypedDict):
    owner_id: str
    amount: Decimal


class RoyaltyBreakdown(TypedDict):
    gross: Decimal
    platform_fee: Decimal
    net: Decimal
    payouts: list[Payout]


def format_ledger_amount(amount: Decimal) -> str:
    """Fixed-point string, exactly 4 decimal places (matches amount.value)."""
    return str(amount.quantize(LEDGER_PLACES))


def distribute_splits(pool: Decimal, splits: Iterable[Split]) -> list[Payout]:
    """
    Floor each owner's share of the pool to the cent, then hand out the
    leftover cents one at a time by largest fractional remainder. Ties
    break by larger bps, then lexicographically smaller owner_id —
    deterministic, same event always yields the same cents. Splits that
    divide evenly (e.g. 60/40 of $1.25) leave nothing to distribute.
    """
    splits = list(splits)
    total_bps = sum(s["bps"] for s in splits)
    if total_bps != 10000:
        raise ValueError(f"splits[].bps must sum to exactly 10000, got {total_bps}")

    raw_shares = [
        (s["owner_id"], s["bps"], pool * Decimal(s["bps"]) / BPS_DENOMINATOR)
        for s in splits
    ]
    floored = {
        owner_id: raw.quantize(CENT, rounding=ROUND_DOWN)
        for owner_id, _, raw in raw_shares
    }
    remainders = {owner_id: raw - floored[owner_id] for owner_id, _, raw in raw_shares}
    bps_by_owner = {owner_id: bps for owner_id, bps, _ in raw_shares}

    pool_cents = int((pool / CENT).to_integral_value(rounding=ROUND_DOWN))
    floor_cents = int(sum(floored.values()) / CENT)
    leftover_cents = pool_cents - floor_cents

    ordering = sorted(
        (s["owner_id"] for s in splits),
        key=lambda owner_id: (-remainders[owner_id], -bps_by_owner[owner_id], owner_id),
    )
    for owner_id in ordering[:leftover_cents]:
        floored[owner_id] += CENT

    return [{"owner_id": s["owner_id"], "amount": floored[s["owner_id"]]} for s in splits]


def compute_royalty(gross: Decimal, splits: Iterable[Split]) -> RoyaltyBreakdown:
    """
    Full breakdown for one royalty obligation. `gross` is the creator
    payout pool in full — no fee is deducted, `net` equals `gross`, and
    `platform_fee` is always zero (kept in the return shape for schema
    stability, not because any fee applies). Raises AssertionError if
    the balanced-journal invariant (sum(payouts) == pool) fails — that
    must never happen for valid inputs and should hard-fail before any
    ledger write is attempted.
    """
    pool = gross.quantize(CENT)
    payouts = distribute_splits(pool, splits)

    total_payout = sum((p["amount"] for p in payouts), Decimal("0"))
    if total_payout != pool:
        raise AssertionError(
            f"balanced-journal invariant violated: payouts({total_payout}) != pool({pool})"
        )

    return {"gross": gross, "platform_fee": Decimal("0.00"), "net": gross, "payouts": payouts}
